File: src/WirePact.PoC.Translator/Interceptors/TimeLogger.cs
using System.Diagnostics;
using System.Threading.Tasks;
using Grpc.Core;
using Grpc.Core.Interceptors;
using Microsoft.Extensions.Logging;

namespace WirePact.PoC.Translator.Interceptors
{
    public class TimeLogger : Interceptor
    {
        private readonly ILogger<TimeLogger> _logger;

        public TimeLogger(ILogger<TimeLogger> logger)
        {
            _logger = logger;
        }

        public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
            TRequest request,
            ServerCallContext context,
            UnaryServerMethod<TRequest, TResponse> continuation)
        {
            var watch = Stopwatch.StartNew();
            var result = await base.UnaryServerHandler(request, context, continuation);
            watch.Stop();
            _logger.LogInformation($"Call to Authentication Check Handler took {watch.ElapsedMilliseconds}ms.");

            return result;
        }
    }
}

File: src/WirePact.PoC.Translator/TimeLogger.fs
namespace WirePact.PoC.Translator

open System.Diagnostics
open FSharp.Control.Tasks
open Grpc.Core.Interceptors
open Microsoft.Extensions.Logging

type TimeLogger(logger: ILogger<TimeLogger>) =
    inherit Interceptor()

    override _.UnaryServerHandler(request, context, continuation) =
        let resultTask =
            base.UnaryServerHandler(request, context, continuation)

        task {
            let watch = Stopwatch.StartNew()
            let! result = resultTask
            watch.Stop()
            logger.LogInformation $"Call to Authentication Check Handler took {watch.ElapsedMilliseconds}ms."

            return result
        }
